### DISCUSSION TITLE
Java 2일차 서희, 재찬

### DISCUSSION THREAD

**Discussion by LireEruel** [Shortudy-10th/tech-for-developer#26] [질의 응답]
# 자바 2일차

**Comment by LireEruel** [comment]
추상 클래스와 인터페이스의 차이점에 대해서 말씀해주세요.

**Reply by LireEruel** [reply]
- 상속
  - 하나의 클래스는 추상 클래스를 하나만 상속받을 수 있습니다. 하지만 인터페이스는 다중으로 구현할 수 있습니다.
  - 상속받는 클래스는 추상 클래스의 추상 메서드만 구현하면 되지만, 인터페이스는 인터페이스의 모든 메서드를 하위 클래스에서 반드시 구현해야 합니다.
- 생성자
  - 추상 클래스에는 생성자가 있어서 초기화 로직을 포함할 수 있습니다. 
  - 인터페이스는 생성자가 없으므로 초기화 로직을 포함시킬 수 없습니다.
- 상태 유지
  - 추상 클래스는 필드(변수)를 가질 수 있으며, 상태를 유지할 수 있습니다.
  - 인터페이스는 상수가 아닌 필드(변수)를 가질 수 없으믈, 상태를 유지할 수 없습니다.
- 메서드
  - 추상 클래스는 추상메서드와 구현된 메서드 모두 포함할 수 있습니다.     
  - 인터페이스는 구현되지 않은 메서드만 가질 수 있었으나 java 8 부터는 default 메서드와 정적 메서드를 가질 수 있습니다.

**Reply by kjc6735** [reply]
상수를 선언하려고 합니다. 
어떻게 선언하실건가요 ? 그 이유도 설명해주세요 .

**Reply by LireEruel** [reply]
@kjc6735 final키워드를 사용한다. 그래야 수정이 되지 않기 떄문이다.

**Reply by kjc6735** [reply]
좀 더 정확하게 하자면
interface나 abstract의 static final키워드를 사용해 선언해야한다.

그 이유는 접근 o, 인스턴스 생성x, 수정 불가해야한다.
그러므로 이것을 만족하려면
abstract의 static final 변수
interface의 static final 변수
둘 중 하나여야 한다.

**Comment by LireEruel** [comment]
오버라이딩과 오버로딩에 대해 설명해주세요.

**Reply by LireEruel** [reply]
오버라이딩 : 부모에게서 상속받은 메서드를 자식이 재정의 하는 것을 말합니다.
오버로딩 : 같은 이름을 가졌지만 다른 파라미터와 다른 기능을 수행하는 함수를 여러개 만드는 것을 말합니다.

**Comment by LireEruel** [comment]
String, StringBuilder, StringBuffer에 대해 설명해주세요.

**Reply by LireEruel** [reply]
String은 reference 타입 객체인 문자열 클래스입니다.
reference 중에서 유일하게 premitive 타입처럼 new 없이 객체를 생성할 수 있다는 특징이 있습니다.
불변성을 가졌기 때문에 문자열을 수정할 때마다 새로운 String 객체가 생성됩니다.
StringBuilder는 문자열을 변경할 수 있는 클래스로 문자열을 조작할 때마다 새로운 객체를 생성하지 않습니다.
StringBuffer는 StringBuilder처럼 문자열을 변경할 수 있는 클래스입니다. 하지만 StringBuilder와는 달리 모든 메서드가 동기화되어 있어 멀티스레드 환경에서도 안전합니다.

### 사용 시 고려 사항
- 문자열 변경 빈도가 낮고 멀티스레드 환경: String 사용
- 문자열 변경 빈도가 높고 단일 스레드 환경: StringBuilder 사용
- 문자열 변경 빈도가 높고 멀티스레드 환경: StringBuffer 사용

**Reply by kjc6735** [reply]
object 의 equals와 == 와 String의 equals의 차이를 설명해주세요

**Reply by kjc6735** [reply]
String의 경우 hashCode가 같다면 equals한가요 ?
반대로 equals 가 true라면 hashcode가 같나요?

**Reply by LireEruel** [reply]
@kjc6735 object 의 equals와 == 와 String의 equals의 차이를 설명해주세요
object와 String 모두 == 은 두 참조가 같은 메모리 주소값을 가리키는지 비교합니다. 이 것을 '동일성 비교'라고 합니다.
object의 equals는 ==과 같이 동일성 비교를 합니다. == 과 달리 오버라이드 하여 동등성 비교를 할 수도 있습니다.
String의 equals는 두 문자열이 동일한 값을 가지고 있는지 확인합니다. 메모리상 위치가 달라도 내용이 같으면 'true'를 반환합니다. 이것을 '동등성 비교' 라고 합니다.

**Reply by LireEruel** [reply]
@kjc6735 

>  String의 경우 hashCode가 같다면 equals한가요 ?반대로 equals 가 true라면 hashcode가 같나요?

hashCode가 같다면 같은 객체라는 뜻 이기 때문에 반드시 equals 합니다. 하지만 String 의 equals는 동일성 비교가 아닌 동등성 비교이기 때문에 equals하다고 해서 같은 hashcode를 갖는 것은 아닙니다.

**[사전 지식]**
Hascode : 객체를 식별하기 위해 사용되는 정수 값. Java에서 hashCode 메서드는 Object 클래스에 정의되어 있으며, 객체의 메모리 주소나 객체의 상태 등을 기반으로 그 객체에 대한 정수 형태의 해시 코드를 생성합니다.

**Reply by kjc6735** [reply]
해쉬코드가 같다고 해서 같은 객체가 아닙니다. 해쉬코드가 같은 경우에는 같은 문자열일 경우도 있지만 해쉬충돌 난 경우에 다른 문자열일 경우 해쉬코드가 같을 수 있습니다. 이 경우에는 해쉬코드가 같아도 equals 하지 않다..

예를 들면 자바1버전에서는 문자열의 앞 16글자를 가지고 해쉬코드를 생성했다.
도메인이 
http://www.test.naver.com/?abc=1&test=2
http://www.test.naver.com/?abc=154&test=3
라고 가정해보면
얖의 16글자로만 해쉬코드를 생성했으니 두 도메인의 해쉬코드는 같게 생성된다.

하지만 문자열은 다르다.
hashcode상으로는 같지만 equals하지 않은 상황이다.

반대로 equals 하면 hashcode가 같을까?
라는 답변은 O이다.

해쉬코드는 내용을 기반으로 해쉬코드를 생성한다. String의 equals는 값을 비교하므로(문자열 하나하나 비교함) 값이 같다면 hashcode는 같다.

정리하자면 hashCode가 같다고 해서 같은 객체가 아니다. 문자열 뿐만이 아니라 다른 오브젝트도 마찬가지로 해쉬코드는 내용을 가지고 해쉬코드를 생성한다. 해쉬코드는 범위가 정해져있기 떄문에 내용이 달라도 해쉬코드 값이 같을 수 있다. 반대로 equals하면 hashcode는 같다. 오브젝트의 내용을 기반으로 해쉬코드를 생성하기 같은 문자열이라면 같은 해쉬코드를 가진다.  

자바에서의 스트링의 동일성과 동등성
자바에서 동일성은 같은 오브젝트인지를 확인하는 것이고, 동등성 비교는 내용을 비교하는 것이다. 
String str = "~~~" 처럼 생성했을 시 스트링풀에 저장된다.
스트링풀에 저장된 문자열은 유일하다. 또한 같은 문자열을 가진 변수가 있을 경우 같은 주소값을 가리킨다.
String의 ==은 동일성 비교이다. 그렇기 때문에 자바에서 같은 문자열일 경우에는 ==비교가 가능하다.(중요 : String str = "~~~" 처럼 생성했을 시)
만약 new Stringa("~~~")처럼 생성했다면 결과는 달라진다. new String은 힙메모리에 올라가며, 스트링풀에 저장되는 객체가 아닌 힙메모리에 올라가는 객체가 된다. 그럼 같은 문자열이라도 ==(동일성)비교에서 false가 나온다(서로의 주소가 다름). 
String의 equals의 경우에는 같은 주소를 갖는지  확인하고, 같다면 문자열 내용을 확인한다. 
 String str=  "test" ; String str1 = new String("test");
 일 경우에 서로 다른 오브젝트이므로 내용을 비교하고 내용이 같으니 true를 리턴한다.

**Comment by LireEruel** [comment]
자바에서 스레드를 사용하기 위한 두 가지 방법에 대해 설명하고 차이점을 알려주세요.

**Reply by LireEruel** [reply]
## 1. ```Thread``` 클래스 확장
자바에서 스레드를 생성하기 위해 ```Thread``` 클래스를 직접 확장할 수 있습니다.
## 2. ```Runnable```인터페이스 구현
스레드를 생성하는 또 다른 방법은 ```Runnable```인터페이스를 구현하는 것입니다.

# 차이점
## 상속 
```Thread``` 클래스를 확장하면 다른 클래스를 상속받을 수 없으므로, 상속이 필요한 경우에는 ```Runnable```인터페이스를 사용하는 것이 좋습니다.
## 구성
```Runnable```인터페이스를 사용하면 스레드 작업을 객체로 분리하여 여러 스레드에서 재사용할 수 있으며, 코드의 결합도를 낮추고 유연성을 높일 수 있습니다.
## 결합도
```Runnable```을 사용하면 스레드와 작업을 분리할 수 있으므로, 다른 스레드에서 같은 작업 인스턴스를 실행할 수 있습니다. 이로 인해 코드의 결합도가 더 낮아집니다.
# 결론
- ```Thread``` 클래스를 확장하는 것은 단순하고 직관적입니다. 하지만 다른 클래스를 상속받아야 하는 경우나 코드의 재사용과 유연성이 중요한 경우에는 적합하지 않을 수 있습니다.
- ```Runnable``` 인터페이스를 사용하는 것은 더 유연하며, 코드의 재사용성과 유지보수성을 높일 수 있습니다. 다양한 상황에서 스레드 작업을 수행하려는 경우, 이 방법이 더 적합할 수 있습니다.

**Comment by kjc6735** [comment]
추상 클래스와 인터페이스의 차이점에 대해서 말씀해주세요.

**Reply by kjc6735** [reply]
추상클래스는 맴버변수를 가질 수 있고, 구현 메서드를 가질 수 있다. 다중 상속이 불가능하다.
인터페이스의 경우에는 추상 메서드만을 가지고 있으며, 구현 메서드를 가질 수 없다. 하지만 자바8로 올라오면서 default 키워드를 사용하는 구현 메서드를 가질 수 있지만, 꼭 구현하지 않아도 되는 메서드이다. 다중 상속이 가능하며, 만약에 변수를 사용한다면, 무조건 static final로 사용해야 한다. 다중상속이 가능하다.

**Reply by LireEruel** [reply]
추상 클래스와 인터페이스는 각각 언제 사용해야할까요?

**Reply by kjc6735** [reply]
일부 메서드를 궁유할 때 사용하고, 인터페이스는 다중상속이나 특정 동작 구현을 강제할 때 사용한다.

**Comment by kjc6735** [comment]
오버라이딩과 오버로딩에 대해 설명해주세요.

**Reply by kjc6735** [reply]
오버라이딩은 부모 클래스에서 정의된 메서드를 재정의 하는 것이다. 
오버로딩은 클래스 내에서 같은 네이밍을 사용하는 메서드를 말한다.
차이점은 
오버라이딩은 매개변수, 타입이 같아야 한다.
오버로딩은 이름만 같고 이름만 같고, 변수 타입, 갯수, 순서 중 하나라도 달라야한다. 
같은 점은 이름이 같아야 한다는 것이다.

**Comment by kjc6735** [comment]
String, StringBuilder, StringBuffer에 대해 설명해주세요.

String : 불변성이기 때문에 자동적으로 thread-safe하다. 만약에 문자열을 수정한다면 새로운 객체로 만들어 반환된다. 문자열을 변경하면 ( ex. String str = "test"; str = "abc") 새로운 객체로 만들며, 만약 문자열이 같다면 == 로 비교가 가능하다. 새로운 객체가 만들어지므로 문자열을 바꿀 때 마다 주소가 바뀐다, 하지만 해쉬충돌로 인해서 다른 문자열이라도 ==가 true가 나오는 경우가 있으므로, equals를 사용하는 것을 권장한다. 
StringBuilder : 가변적이며, 문자열을 수정할 수 있음. 하지만 쓰레드 세이프하지 않으므로 싱글스레드에서 사용하는 것이 좋다. 문자열을 변경해도 주소값은 바뀌지 않으며, 힙메모리에 올라간다.  
StringBuffer : 스트링빌더와 같음. 다른 점은 내부적으로 동기화 처리가 되어있어 멀티 스레드 환경에서 사용해도 안전하다.

**Comment by kjc6735** [comment]
자바에서 스레드를 사용하기 위한 두 가지 방법에 대해 설명하고 차이점을 알려주세요.

**Reply by kjc6735** [reply]
쓰레드를 사용하기 위한 방법은 2가지가 있다.
thread 클래스를 사용하는 것과 runnable을 구현하는 것이다. 
thead는 구현체이며, runnable은 인터페이스이다. 그렇기 때문에 쓰레드는 다중 상속이 불가능하고, Runnable은 다중 상속이 가능하게 구현할 수 있다.

**Reply by LireEruel** [reply]
runnable 과 thread 은 각각 언제 사용하나요?

**Reply by kjc6735** [reply]
이미 다른 클래스를 상속받았다면 interface인 runnable을 사용하고, 기존 클래스를 쓰레드를 적용하려면 쓰레드를 상속받는다.

**Comment by kjc6735** [comment]
7. 객체 지향 프로그래밍의 4가지 특징에 대해서 말씀해주세요.

**Reply by kjc6735** [reply]
캡슐화 : 데이터를 직접 접근하는 것을 막으며, 정보 은닉이 가능하게 해준다. 
상속 : 이미 존재하는 클래스로부터 상속받아 새로운 클래스를 만드는 것이다.  코드의 재사용성을 높일 수다.
다형성 : 같은 이름의 메서드를 각 객체에서 다시 정의해, 서로 다른 클래스의 객체들이 동일한 메서드를 호출 할 수 있다.
추상화 : 객체의 공통된 특성과 동작을 정의하는 것으로 추상 클래스나 인터페이스를 통해 추상화랄 표현할 수 있다.

**Reply by LireEruel** [reply]
캡슐화 (Encapsulation)
캡슐화는 객체의 상태와 행동을 하나로 묶고, 객체의 상태를 외부에서 직접 접근할 수 없게 하여 내부 구현을 숨기는 것을 말합니다. 이를 통해 객체의 내부 상태를 보호하고, 외부에서는 해당 객체의 메서드를 통해서만 상태를 변경할 수 있도록 제어합니다.

상속 (Inheritance)
상속은 한 클래스의 특성(속성과 메서드)을 다른 클래스가 물려받을 수 있게 해줍니다. 상속을 통해 코드의 재사용성을 높이고 중복을 최소화할 수 있으며, 기반 클래스의 변경이 파생 클래스에 자동으로 반영되므로 유지보수가 용이해집니다.

다형성 (Polymorphism)
다형성은 하나의 인터페이스나 부모 클래스를 공유하는 여러 클래스의 객체가 동일한 메서드 호출에 대해 각각 다르게 반응할 수 있게 하는 특징입니다. 이를 통해 코드의 유연성과 확장성을 증가시키며, 클래스 간의 결합도를 낮출 수 있습니다.

추상화 (Abstraction)
추상화는 객체의 특징 중 중요한 부분에만 집중하고 불필요한 세부 사항은 생략하는 것을 말합니다. 추상화를 통해 시스템의 복잡도를 줄이고 유지보수를 용이하게 할 수 있습니다.
